### PR TITLE
Migrate Kubernetes StatefulSet to Advanced StatefulSet automatically

### LINKED ISSUE #1115: advanced-statefulset: upgrade to advanced-statefulset automatically
## Feature Request

**Is your feature request related to a problem? Please describe:**

Existing pods must be adopted by Advanced Statefulset.

**Describe the feature you'd like:**

Able to upgrade the tidb-operator to use advanced stateful set.

Note that it's not possible to revert it back automatically.

https://github.com/pingcap/advanced-statefulset/blob/ef3ba0af1714bacc8e4175a6443e0cd226ea2d84/test/e2e/apps/pingcap.go#L309-L447

### PR DESCRIPTION
**Pull request by cofyc**
Signed-off-by: Yecheng Fu <fuyecheng@pingcap.com>

### What problem does this PR solve? 

fixes #1115 

### Check List 

Tests 

 - Unit test
 - E2E test
 - Stability test
 - Manual test (add detailed scripts or steps below)
 - No code

Code changes

 - Has Go code change
 - Has CI related scripts change
 - Has Terraform scripts change

Side effects

 - Breaking backward compatibility

Related changes

 - Need to cherry-pick to the release branch
 - Need to update the documentation

### Does this PR introduce a user-facing change?:

```release-note
Able to migrate existing Kubernetes StatefulSets to Advanced StatefulSets automatically when AdvancedStatfulSet feature is enabled
```

### CODE REVIEW

File: tests/e2e/tidbcluster/tidbcluster.go [hunk 1]
```diff
@@ -704,7 +704,7 @@ var _ = ginkgo.Describe("[tidb-operator] TiDBCluster", func() {
 	})
 
 	ginkgo.It("should be operable without helm [API]", func() {
-		tc := fixture.GetTidbCluster(ns, "plain-cr", "v2.1.16")
+		tc := fixture.GetTidbCluster(ns, "plain-cr", utilimage.TiDBV2Version)
```
Comment (cofyc) [inline tests/e2e/tidbcluster/tidbcluster.go]: use shared version to reduce the total images we used in e2e

File: pkg/upgrader/upgrader.go [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+			klog.Infof("Upgrader: found 0 Kubernetes StatefulSets, nothing need to do")
+			return nil
+		}
+		klog.Infof("Upgrader: found %d Kubernetes StatefulSets, trying to migrate one by one", len(stsList.Items))
+		for _, sts := range stsList.Items {
+			_, err := helper.Upgrade(u.kubeCli, u.asCli, &sts)
+			if err != nil {
+				return err
+			}
+			klog.Infof("Upgrader: successfully migrated Kubernetes StatefulSet %s/%s", sts.Namespace, sts.Name)
```
Comment (aylei) [inline pkg/upgrader/upgrader.go]: I think we should only migrate statefulset owned by tidbcluster

File: tests/e2e/tidbcluster/serial.go [hunk 3]
```diff
@@ -297,6 +300,97 @@ var _ = ginkgo.Describe("[tidb-operator][Serial]", func() {
...
+		var err error
+		genericCli, err = client.New(config, client.Options{Scheme: scheme.Scheme})
+		framework.ExpectNoError(err, "failed to create clientset")
+
+		defer func() {
+			// ginkgo.By("Uninstall tidb-operator")
+			// oa.CleanOperatorOrDie(ocfg)
+			// ginkgo.By("Uninstalling CRDs")
+			// oa.CleanCRDOrDie()
+		}()
```
[the remark this answers is not in the document]
Comment (cofyc) [reply tests/e2e/tidbcluster/serial.go]: ah, I should uncomment these lines.

File: pkg/upgrader/upgrader.go (hunk 2 above)
Comment (cofyc) [reply pkg/upgrader/upgrader.go]: good catch!

File: tests/e2e/tidbcluster/serial.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (cofyc) [reply tests/e2e/tidbcluster/serial.go]: done

File: pkg/upgrader/upgrader.go (hunk 2 above)
Comment (cofyc) [reply pkg/upgrader/upgrader.go]: fixed in https://github.com/pingcap/tidb-operator/pull/1580/commits/d9f0d8ed5b069f13e609caa8916e5b37db7bd3d8

**Comment by Yisaer** [comment]
@cofyc  Could add some details to explain how `helper.Upgrade` work? I think it's the most important information in this request.

**Comment by Yisaer** [comment]
> @cofyc Could add some details to explain how `helper.Upgrade` work? I think it's the most important information in this request.

The Rest LGTM

**Comment by cofyc** [comment]
sure, the basic procedure is like this:

- remove sts selector labels from controller revisions and set a special annotation for Advanced StatefulSet (can be skipped if Kubernetes cluster has http://issues.k8s.io/84982 fixed)
- create advanced sts
- delete sts with DeletePropagationOrphan policy

https://github.com/pingcap/advanced-statefulset/blob/087fc829087b081c91e05af323d82fe4849c510e/pkg/apis/apps/v1/helper/upgrade.go#L32-L116

**Pull request merged**